### PR TITLE
V2/refactor/models

### PR DESCRIPTION
**Pull request by davidegraff**
## Description
This PR refactors large portions of the underlying model code focusing on composition and protocol usage rather than inheritance

## Current model and limitations
Currently, an `MPNN` is an abstract `LightningModule` that accepts a `MessagePassingBlock` and a bunch of arguments related to FFN construction and training dynamics. However, to actually use an `MPNN`, a client must instead instantiate one of the task/target-specific child classes: `RegressionMPNN`, `ClassificationMPNN`, etc. Each child class handles not only output transformation (reshaping for multiclass, softplus activation at prediction for MVE variances, etc.) but also controls the selection of _valid_ losses/metrics for the class. Altogether, the code as written, is generally messy, relies too heavily on inheritance, and is not flexible enough for future extension.

## Summary of major changes
- **separate message passing and readout into separate modules**: `MessagePassingBlock` and `Aggregation`
As the code is currently written, it's pretty challenging for a client to access the node-level hidden representations prior to graph aggregation without hacking something. This will be necessary for a future reimplementation of atom- and bond-level property prediction. Moreover, clients can now define their own aggregation scheme as opposed to being limited to the ones we've implemented.
- **abstracting out property readout from the MPNN**
the current MPNN accepts a bunch of FFN-related arguments that then get passed into the `build_ffn` function. But why should we limit clients to readout networks that can be built from that function? If a client wants to access a more exotic FFN architecture, they would need to subclass `MPNN`, but that's overkill when we could just let composition do the job for us.
- **a larger focus on the use of `Protocol`s**
It's always possible to define a base implementation and have clients just inherit and override what they want to, but at that point, you're mixing together an interface and an implementation. This PR has started to abstract these two apart.

## Questions
I've spent a good amount of time perusing PyTorch Lightning discussion forums, but I haven't found a satisfying answer to the following questions:
- handling variable loss functions in your `pl.LightningModule`. These are usually hardcoded into the module itself, but we have a few options for some of our downstream tasks
- a similar thing for evaluation metrics, especially as they relate to monitoring validation loss for early stopping, checkpointing, etc.
- model saving/loading. We could probably just pickle the whole thing, but I'd like a longer-term solution that can decouple the _architecture_ from the _parameters_

## Checklist
- [x] linted with flake8?
- [ ] (if appropriate) unit tests added?

### CODE REVIEW

File: chemprop/v2/cli/train.py [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+                p_bond_feats=args.val_bond_features_path,
+                p_atom_descs=args.val_atom_descriptors_path,
+                **format_kwargs,
+                **featurization_kwargs,
+            )
+            train_data = all_data
+        else:
+            train_data, val_data, _ = split_data(all_data, args.split, args.split_sizes)
+    else:
+        raise ArgumentError("'val_path' must be specified is 'test_path' is provided!")
```
Comment (KnathanM) [inline chemprop/v2/cli/train.py:297]: I think this should be:  ... must be specified if ...
Making a comment so that me or someone else will remember to fix that later.
